contracts/CompoundV3InvestStrategy.sol

```diff
@@ -0,0 +1,148 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity 0.8.16;
+
+import {ICompoundV3} from "./interfaces/ICompoundV3.sol";
+import {ICometRewards} from "./interfaces/ICometRewards.sol";
+import {IERC20} from "@openzeppelin/contracts/interfaces/IERC20.sol";
+import {IAccessControl} from "@openzeppelin/contracts/access/IAccessControl.sol";
+import {SwapLibrary} from "@ensuro/swaplibrary/contracts/SwapLibrary.sol";
+import {IInvestStrategy} from "./interfaces/IInvestStrategy.sol";
+import {StorageSlot} from "@openzeppelin/contracts/utils/StorageSlot.sol";
+import {IExposeStorage} from "./interfaces/IExposeStorage.sol";
+import {InvestStrategyClient} from "./InvestStrategyClient.sol";
+
+/**
+ * @title CompoundV3ERC4626
+ * @dev Vault that invests/deinvests into CompoundV3 on each deposit/withdraw. Also, has a method to claim the rewards,
+ *      swap them, and reinvests the result into CompoundV3.
+ *      Entering or exiting the vault is permissioned, requires LP_ROLE
+ *
+ * @custom:security-contact security@ensuro.co
+ * @author Ensuro
+ */
+contract CompoundV3InvestStrategy is IInvestStrategy {
+  using SwapLibrary for SwapLibrary.SwapConfig;
+
+  bytes32 public constant HARVEST_ROLE = keccak256("HARVEST_ROLE");
+  bytes32 public constant SWAP_ADMIN_ROLE = keccak256("SWAP_ADMIN_ROLE");
+
+  address private immutable __self = address(this);
+  bytes32 public immutable storageSlot = InvestStrategyClient.makeStorageSlot(this);
+
+  /// @custom:oz-upgrades-unsafe-allow state-variable-immutable
+  ICompoundV3 internal immutable _cToken;
+  /// @custom:oz-upgrades-unsafe-allow state-variable-immutable
+  ICometRewards internal immutable _rewardsManager;
+  address internal immutable _baseToken;
+
+  event RewardsClaimed(address token, uint256 rewards, uint256 receivedInAsset);
+
+  event SwapConfigChanged(SwapLibrary.SwapConfig oldConfig, SwapLibrary.SwapConfig newConfig);
+
+  // From OZ v5
+  error AccessControlUnauthorizedAccount(address account, bytes32 neededRole);
+  error CanBeCalledOnlyThroughDelegateCall();
+  error CannotDisconnectWithAssets();
+
+  enum ForwardMethods {
+    harvestRewards,
+    setSwapConfig
+  }
+
+  modifier onlyDelegCall() {
+    if (address(this) == __self) revert CanBeCalledOnlyThroughDelegateCall();
+    _;
+  }
+
+  modifier onlyRole(bytes32 role) {
+    if (!IAccessControl(address(this)).hasRole(role, msg.sender))
+      revert AccessControlUnauthorizedAccount(msg.sender, role);
+    _;
+  }
+
+  constructor(ICompoundV3 cToken_, ICometRewards rewardsManager_) {
+    _cToken = cToken_;
+    _rewardsManager = rewardsManager_;
+    _baseToken = cToken_.baseToken();
+  }
+
+  function connect(bytes memory initData) external virtual override onlyDelegCall {
+    SwapLibrary.SwapConfig memory swapConfig = abi.decode(initData, (SwapLibrary.SwapConfig));
+    swapConfig.validate();
+    StorageSlot.getBytesSlot(storageSlot).value = initData;
```

Comment:
Should this be reencoded from the validated swapconfig to make sure there's no additional data?